Add initial implementation of `std::dump` (#1)
bemanproject/dump: README.md

SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
-->

> # beman.dump: Implementation of proposed `std::dump`

Reviewer: ```suggestion
# beman.dump: A tool for dumping an object's value to standard output
```

IMO it's better, from a user perspective, to describe the functionality this contains instead of forcing someone to look elsewhere.

![Continuous Integration Tests](https://github.com/beman-project/dump/actions/workflows/ci_tests.yml/badge.svg)

> `std::dump` prints its arguments space-separated with a new-line.

Reviewer: ```suggestion
`beman::dump::dump()` prints its arguments space-separated with a new-line. This
is useful as a debugging utility.
```

It's important to explain to users when an why they would use this library; what it's purpose is.

> A call to `std::dump(arg1, arg2, …, argn)` is equivalent to `std::println(“{} {} … {}”, arg1, arg2, …, argn)`

Reviewer: ```suggestion
A call to `beman::dump::dump(arg1, arg2, …, argn)` is equivalent to `std::println(“{} {} … {}”, arg1, arg2, …, argn)`
```

> `std::dump` is proposed by [P2879](https://wg21.link/P2879). This project currently implements R0.

Reviewer: ```suggestion
- Implements: [Proposal of `std::dump` (P2879R0)](https://wg21.link/P2879R0)
```

> ## Building beman.dump

Reviewer: There should be a `## Usage` section demonstrating `dump()` calls.

Reviewer 2: Added TODO

</details>

### How to build beman.dump

This project strives to be as normal and simple a CMake project as possible. This build workflow in particular will work, producing a static `libbeman.dump.a` library, ready to package with its headers:

```shell
cmake -B build -S .
cmake --build build
ctest --test-dir build
cmake --install build --prefix /opt/beman.dump
```
## Integrate beman.dump into your project

<details>
<summary> Use beman.dump directly from C++ </summary>
<!-- TODO Darius: rewrite section!-->

If you want to use `beman.dump` from your project, you can include `beman/dump/*.hpp`  files from your C++ source files

```cpp

> #include <beman/dump/identity.hpp>

Reviewer: ```suggestion
#include <beman/dump/dump.hpp>
```
